# 1 Prepare your environment

In [1]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import random as rnd

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# machine learning
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

NOTE: If you want to try additional algorithms, you might also add them from scikitlearn: 

In [2]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier

from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier

# 2 Load datasets

Downoad the Titanic datasets from Kaggle (you can also take them from the campus). We will use panda for organizing and working with data. We will load training and test data separately in two dataframes, and we will create a combined dataset by merging both. 

For the practical session, we will use only the train.csv file as it contains labeled data. 
The test.csvfile is thought to be used in the challenge.

NOTE: You need to upload the CSV files into Colab before running these lines.

In [4]:
titanic_df = pd.read_csv('train.csv')

*Q1. Which features are available in the dataset?*

Com podem veure en la següent línia de codi, els features del dataset són els seguents:

Passenger Id', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked.

Els hem obtingut amb la funció d'imprimir els noms de les columnes interada a panda.

In [6]:
# you can see the features by printing the column names using panda built functions

# WRITE YOUR SOLUTION HERE: 
print (titanic_df.columns.values) 

['PassengerId' 'Survived' 'Pclass' 'Name' 'Sex' 'Age' 'SibSp' 'Parch'
 'Ticket' 'Fare' 'Cabin' 'Embarked']


*Q2. Which features are categorical?*

Observe the list of features and list the ones thar are categorical (i.e. they classify the samples into sets). These are also known as ***qualitative features***.

Les features que són categorical són les següents:

Fare, Embarked, Sex, Name.

*Q3. Which features are numerical?*

Observe the list of features and list the ones thar are numerical (i.e. they have values that change from sample to sample). Numerical values can be discrete, continuous or timeseries based. These are also known as ***quantitative features***.

Quantitavive features are the following:

Passenger ID, Survived, Pclass, SibSp, Parch.

In [7]:
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

*Q4. Which features are mixed data types?*

These contain numerical, alphanumeric data within the same feature. They are candidates for applying correction strategies.

Mixed features are the following:

Cabin and Ticket.

*Q5. Which features may contain errors or typos? Which features contain blanck, null or empty values?*

This is harder to review for a large dataset, however reviewing a few samples from a smaller dataset may give us some cues. For example, we can plot the tail (last 5 rows) of the dataset.

Features with errors:

Age, Cabin.

Features with empty values:

Survived, SibSp, Parch

In [8]:
titanic_df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

*Q7. What are the data types for various features?*

This will help us during converting and for maipulating the diferent data columns. We can use the `info` function from panda.

Data types:

    1. int64
    2. object
    3. float64

In [10]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


*Q8. What is the distribution of numerical feature values across the samples?*

You might also consider this additional information: 
- Total samples are 891 or 40% of the actual number of passengers on board the Titanic (2,224).
- Survived is a categorical feature with 0 or 1 values.
- Around 38% samples survived representative of the actual survival rate at 32%.
- Most passengers (> 75%) did not travel with parents or children.
- Nearly 30% of the passengers had siblings and/or spouse aboard.
- Fares varied significantly with few passengers (<1%) paying as high as $512.
- Few elderly passengers (<1%) within age range 65-80.

You can use the `describe` function from pandas
(see [here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html)) 
to retrieve this information. 